server/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -366,10 +366,34 @@ public T getDefault(Settings settings) {
     }
 
     /**
-     * Returns <code>true</code> iff this setting is present in the given settings object. Otherwise <code>false</code>
+     * Returns true if and only if this setting is present in the given settings instance. Note that fallback settings are excluded.
+     *
+     * @param settings the settings
+     * @return true if the setting is present in the given settings instance, otherwise false
      */
-    public boolean exists(Settings settings) {
-        return settings.keySet().contains(getKey());
+    public boolean exists(final Settings settings) {
+        return exists(settings, false);
+    }
+
+    /**
+     * Returns true if and only if this setting optionally including fallback settings is present in the given settings instance.
+     *
+     * @param settings the settings
+     * @param fallback whether or not to include fallback settings
+     * @return true if the setting and optionally fallback settings is present in the given settings instance, otherwise false
+     */
+    public boolean exists(final Settings settings, final boolean fallback) {
+        Setting<?> current = this;
```

Comment:
I think it'd be easier to read if you pulled the `fallback` case out of the loop.

I wonder if instead of `exists(setting, true)` we should do `settingOrFallbackExists(setting)`. Changing behavior based on method name seems more obvious to me than a boolean parameter. I know you IntelliJ folks get the name of boolean parameters inline but the rest of us don't so we have to either remember or reread the method definition.

Reply:
I agree on the naming suggested here. A differently named method is much clearer than an optional extra argument.

Reply:
@nik9000 I pushed eb9be45445ecfb62a8d20c3de97d5e6aa61efe55. Let me know what you think?